test_scripts/test_post_action_verification.py

```diff
@@ -0,0 +1,228 @@
+import unittest
+from unittest.mock import Mock
+
+from build.lib.puma.state_graph.puma_driver import PumaClickException
+from puma.state_graph.action import action
+from puma.state_graph.puma_driver import PumaDriver
+from puma.state_graph.state import SimpleState, ContextualState, State
+from puma.state_graph.state_graph import StateGraph
+from puma.utils import gtl_logging
+
+
+class MockChatState(SimpleState, ContextualState):
+
+    def __init__(self, parent_state: State):
+        super().__init__(xpaths=['xpath'], parent_state=parent_state)
+
+    def validate_context(self, driver: PumaDriver, conversation: str = None) -> bool:
+        return True
+
+
+class MockApplication(StateGraph):
+    main_state = SimpleState(['xpath'], initial_state=True)
+    settings_state = SimpleState(['xpath'], parent_state=main_state)
+    chat_state = MockChatState(parent_state=main_state)
+
+    main_state.to(settings_state, lambda: print('opening setting screen'))
+    main_state.to(chat_state, lambda: print('opening chat'))
+
+    # don't call super.__init__ so we do not try to connect to a real device
+    def __init__(self, driver=Mock(), gtl_logger=Mock()):
+        self.current_state = self.initial_state
+        self.driver = driver
+        self.gtl_logger = gtl_logger
+
+        self.username = None
+        self.messages = {}
+
+    @action(settings_state)
+    def change_username(self, new_name: str):
+        self.username = new_name
+
+    @action(chat_state)
+    def send_message(self, message: str, conversation: str = None, ):
+        if conversation not in self.messages:
+            self.messages = []
+        self.messages.append(message)
+
+
+class TestVerifyWith(unittest.TestCase):
+
+    def setUp(self):
+        # used to capture values inside passed 'verify_with' functions
+        self.captures = []
+
+    def test_verify_with_callable_is_called(self):
+        # test if the verify_with is called by just capturing a boolean
+        def should_be_called():
+            self.captures.append(True)
+            return True
+
+        application = MockApplication()
+        application.change_username(new_name='Capture this', verify_with=should_be_called)
+
+        self.assertEqual(self.captures[0], True)
+
+    def test_verify_with_not_callable_raises_exception(self):
+        application = MockApplication()
+
+        with self.assertRaisesRegex(TypeError, "'verify_with' must be a callable"):
+            application.change_username(new_name='This is illegal', verify_with='not a callable')
+
+    def test_verify_with_raises_puma_click_exception_is_logged(self):
+        def verification_throws_puma_click_exception():
+            raise PumaClickException('puma click exception')
+
+        mock_logger = gtl_logging.create_gtl_logger('mock_udid')
+
+        with self.assertLogs(mock_logger, level='INFO') as logs:
```

Comment:
~~question: Where does `assertLogs` come from? Additionally, should it be named `assert_logs`?~~